source/ring_buffer.c

```diff
@@ -0,0 +1,206 @@
+            aws_atomic_store_ptr(&ring_buf->head, ring_buf->allocation + requested_size);
+            *dest = aws_byte_buf_from_empty_array(ring_buf->allocation, requested_size);
+            return AWS_OP_SUCCESS;
+        }
+    }
+
+    return aws_raise_error(AWS_ERROR_NO_AVAILABLE_BUFFERS);
+}
+
+int aws_ring_buffer_acquire_up_to(struct aws_ring_buffer *ring_buf, size_t requested_size, struct aws_byte_buf *dest) {
+    if (requested_size == 0) {
+        return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+    }
+
+    uint8_t *tail_cpy = aws_atomic_load_ptr(&ring_buf->tail);
+    uint8_t *head_cpy = aws_atomic_load_ptr(&ring_buf->head);
+
+    /* this branch is, we don't have any vended buffers. */
+    if (head_cpy == tail_cpy) {
+        size_t ring_space = ring_buf->allocation_end - ring_buf->allocation;
+        size_t current_space = ring_buf->allocation_end - head_cpy;
+
+        if (!ring_space) {
+            return aws_raise_error(AWS_ERROR_NO_AVAILABLE_BUFFERS);
+        }
+
+        if (current_space >= requested_size) {
+            aws_atomic_store_ptr(&ring_buf->head, head_cpy + requested_size);
+            *dest = aws_byte_buf_from_empty_array(head_cpy, requested_size);
+            return AWS_OP_SUCCESS;
+        }
+
+        if (ring_space >= requested_size) {
+            aws_atomic_store_ptr(&ring_buf->head, ring_buf->allocation + requested_size);
+            /* we don't have any vended, this should be safe. */
+            aws_atomic_store_ptr(&ring_buf->tail, ring_buf->allocation);
+            *dest = aws_byte_buf_from_empty_array(ring_buf->allocation, requested_size);
+            return AWS_OP_SUCCESS;
+        }
+
+        /* go as big as we can. */
+        /* we don't have any vended, so this should be safe. */
+        aws_atomic_store_ptr(&ring_buf->head, ring_buf->allocation + ring_space);
+        aws_atomic_store_ptr(&ring_buf->tail, ring_buf->allocation);
+        *dest = aws_byte_buf_from_empty_array(ring_buf->allocation, ring_space);
+        return AWS_OP_SUCCESS;
+    }
+    /* you'll constantly bounce between the next two branches as the ring buffer is traversed. */
+    /* after N + 1 wraps */
+    if (tail_cpy > head_cpy) {
+        size_t space = tail_cpy - head_cpy;
+        /* this shouldn't be possible. */
+        AWS_ASSERT(space);
+        space -= 1;
+
+        size_t returnable_size = space > requested_size ? requested_size : space;
+
+        if (returnable_size) {
+            aws_atomic_store_ptr(&ring_buf->head, head_cpy + returnable_size);
+            *dest = aws_byte_buf_from_empty_array(head_cpy, returnable_size);
+            return AWS_OP_SUCCESS;
+        }
+        /* after N wraps */
+    } else if (tail_cpy < head_cpy) {
+        size_t head_space = ring_buf->allocation_end - head_cpy;
+        size_t tail_space = tail_cpy - ring_buf->allocation;
+
+        if (!head_space && tail_space <= 1) {
+            return aws_raise_error(AWS_ERROR_NO_AVAILABLE_BUFFERS);
+        }
```

Comment:
if my suspicion about not needing to leave an empty byte at the end is true, we can alter a bunch of logic in this branch...